platform/api/esx/api.go

```diff
@@ -0,0 +1,759 @@
+	opts = append(opts, &types.OptionValue{
+		Key:   "guestinfo.ovfenv",
+		Value: ovfEnv,
+	})
+
+	return &types.VirtualMachineConfigSpec{
+		ExtraConfig: opts,
+	}
+}
+
+func (a *API) addSerialPort(vm *object.VirtualMachine) error {
+	devices, err := vm.Device(a.ctx)
+	if err != nil {
+		return fmt.Errorf("couldn't get devices for vm: %v", err)
+	}
+
+	d, err := devices.CreateSerialPort()
+	if err != nil {
+		return fmt.Errorf("couldn't create serial port: %v", err)
+	}
+
+	err = vm.AddDevice(a.ctx, d)
+	if err != nil {
+		return fmt.Errorf("couldn't add serial port to vm: %v", err)
+	}
+
+	return a.connectSerialPort(vm)
+}
+
+func (a *API) connectSerialPort(vm *object.VirtualMachine) error {
+	devices, err := vm.Device(a.ctx)
+	if err != nil {
+		return fmt.Errorf("couldn't get devices for vm: %v", err)
+	}
+
+	d, err := devices.FindSerialPort("")
+	if err != nil {
+		return fmt.Errorf("couldn't find serial port for vm: %v", err)
+	}
+
+	var mvm mo.VirtualMachine
+	err = vm.Properties(a.ctx, vm.Reference(), []string{"config.files.logDirectory"}, &mvm)
+	if err != nil {
+		return fmt.Errorf("couldn't get log directory: %v", err)
+	}
+	uri := path.Join(mvm.Config.Files.LogDirectory, "serial.out")
+
+	return vm.EditDevice(a.ctx, devices.ConnectSerialPort(d, uri, false, ""))
+}
+
+func (a *API) GetConsoleOutput(name string) (string, error) {
+	_, _, _, datastore, _, err := a.getServerDefaults()
+	if err != nil {
+		return "", fmt.Errorf("couldn't get server defaults: %v", err)
+	}
+
+	uri := fmt.Sprintf("%s/serial.out", name)
+
+	p := soap.DefaultDownload
+
+	f, _, err := datastore.Download(a.ctx, uri, &p)
+	if err != nil {
+		return "", fmt.Errorf("couldn't download console logs: %v", err)
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(f)
+	return buf.String(), nil
+}
+
+func (a *API) CreateDevice(name string, conf *conf.Conf) (*ESXMachine, error) {
+	ignitionData := base64.StdEncoding.EncodeToString(conf.Bytes())
+
+	finder, datacenter, resourcePool, datastore, network, err := a.getServerDefaults()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get server defaults: %v", err)
+	}
+
+	baseVM, err := finder.VirtualMachine(a.ctx, a.options.BaseVMName)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't find base VM: %v", err)
+	}
+
+	folders, err := datacenter.Folders(a.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting datacenter folders: %v", err)
+	}
+	folder := folders.VmFolder
+
+	cloneSpec, err := a.buildCloneSpec(baseVM, folder, network, resourcePool, datastore, ignitionData)
+
+	task, err := baseVM.Clone(a.ctx, folder, name, *cloneSpec)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't clone base VM: %v", err)
+	}
+
+	err = task.Wait(a.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't clone base VM: %v", err)
```

Comment:
having identical error messages (here and the previous `fmt.Errorf`) becomes annoying when one of them gets hit and you don't know which